File: java/demos/ps/dungeons/Guaron_morgue.java
package demos.ps.dungeons;

import static demos.ps.oo.PSGame.getItem;
import core.Script;
import demos.ps.PSDungeon;
import demos.ps.oo.Enemy;
import demos.ps.oo.PSBattle;
import demos.ps.oo.PSGame;
import demos.ps.oo.PSBattle.BattleOutcome;
import demos.ps.oo.PSGame.Chest;
import demos.ps.oo.PSGame.Flags;
import demos.ps.oo.PSGame.Planet;
import demos.ps.oo.PSGame.Trap;
import demos.ps.oo.PSGame.Trapped;
import demos.ps.oo.PSLibEnemy.PS1Enemy;
import demos.ps.oo.PSLibItem.OriginalItem;
import demos.ps.oo.PSMenu.Scene;

public class Guaron_morgue {
	
	public static void startmap() {
		PSDungeon dungeon = PSGame.currentDungeon;
		dungeon.setRandomEnemies(0, new PS1Enemy[]{PS1Enemy.BATALION, PS1Enemy.ZOMBIE, PS1Enemy.REVENANT});
		dungeon.setRandomEnemies(-1, new PS1Enemy[]{PS1Enemy.BATALION, PS1Enemy.ZOMBIE, PS1Enemy.REVENANT});

		dungeon.setFixedEnemies(0, new PS1Enemy[]{PS1Enemy.BATALION, PS1Enemy.REVENANT});
		dungeon.setFixedEnemies(-1, new PS1Enemy[]{PS1Enemy.ZOMBIE, PS1Enemy.REVENANT, PS1Enemy.BATALION});

		dungeon.startDungeon();
	}
	
	public static void chest1() {
		BattleOutcome outcome = BattleOutcome.WIN;
		if(!PSGame.gameData.chestFlags.contains(Chest.GUARON_MORGUE_CHEST1)) {
			PSBattle battle = new PSBattle();
			outcome = battle.battleScene(Scene.CORRIDOR, new Enemy[]{PSGame.getEnemy(PS1Enemy.DEATH_KNIGHT)});
		}
		if(outcome == BattleOutcome.WIN) {
			PSGame.chestFlag(Chest.GUARON_MORGUE_CHEST1, 0, Trapped.NO_TRAP, getItem(OriginalItem.Armor_Laconian_Armor));
		}
		
	}
	public static void chest2() {
		PSGame.chestFlag(Chest.GUARON_MORGUE_CHEST2, 0, Trapped.NO_TRAP, getItem(OriginalItem.Inventory_Dimate));
	}

	public static void stairs1_up() {
		PSGame.warp(5, 7, false);
	}
	public static void stairs1_down() {
		PSGame.warp(21, 10, false);
	}
	public static void stairs2_up() {
		PSGame.warp(12, 12, false);
	}
	public static void stairs2_down() {
		PSGame.warp(25, 13, false);
	}
	public static void stairs3_up() {
		PSGame.warp(9, 3, false);
	}
	public static void stairs3_down() {
		PSGame.warp(25, 6, false);
	}
	public static void stairs4_up() {
		PSGame.warp(13, 7, false);
	}
	public static void stairs4_down() {
		PSGame.warp(27, 8, false);
	}

	public static void trap1() {
		PSGame.trapRoutine(Trap.GUARON_TRAP1, Trap.INFO_GUARON_TRAP1, 25, 8);
	}
	public static void trap2() {
		PSGame.trapRoutine(Trap.GUARON_TRAP2, Trap.INFO_GUARON_TRAP2, 21, 12);
	}

	public static void exit() {
		PSGame.mapswitch(Planet.DEZORIS,80,7);
	}
}

File: src/demos/ps/dungeons/Guaron_morgue.ts
/**
 * Guaron_morgue - Dungeon Script
 * TypeScript port of Guaron_morgue.java
 */

import { PSGame } from '../PSGame';
import { Chest, Trapped, Trap } from '../game/GameData';
import { Planet } from '../game/City';
import { OriginalItem } from '../game/PSLibItem';
import { PSLibEnemy, PS1Enemy } from '../game/PSLibEnemy';
import { PSSceneType } from '../PSMenu';
import { BattleOutcome, PSBattle } from '../battle/PSBattle';
import { MainEngine } from '../../../core/MainEngine';

export class Guaron_morgue {

  public static async startmap(): Promise<void> {
    MainEngine.setScriptContext(Guaron_morgue);
    const dungeon = PSGame.getCurrentDungeonInstance();
    if (dungeon) {
      dungeon.setRandomEnemies(0, [PS1Enemy.BATALION, PS1Enemy.ZOMBIE, PS1Enemy.REVENANT]);
      dungeon.setRandomEnemies(-1, [PS1Enemy.BATALION, PS1Enemy.ZOMBIE, PS1Enemy.REVENANT]);

      dungeon.setFixedEnemies(0, [PS1Enemy.BATALION, PS1Enemy.REVENANT]);
      dungeon.setFixedEnemies(-1, [PS1Enemy.ZOMBIE, PS1Enemy.REVENANT, PS1Enemy.BATALION]);

      await dungeon.startDungeon();
    }
  }

  public static async chest1(): Promise<void> {
    let outcome = BattleOutcome.WIN;
    if (!PSGame.gameData.chestFlags.has(Chest.GUARON_MORGUE_CHEST1)) {
      const battle = new PSBattle();
      outcome = await battle.battleSceneWithEnemies(PSSceneType.CORRIDOR, [PSLibEnemy.getEnemyByEnum(PS1Enemy.DEATH_KNIGHT)!]);
    }
    if (outcome === BattleOutcome.WIN) {
      await PSGame.chestFlag(Chest.GUARON_MORGUE_CHEST1, 0, Trapped.NO_TRAP, PSGame.getItem(OriginalItem.Armor_Laconian_Armor));
    }
  }

  public static async chest2(): Promise<void> {
    await PSGame.chestFlag(Chest.GUARON_MORGUE_CHEST2, 0, Trapped.NO_TRAP, PSGame.getItem(OriginalItem.Inventory_Dimate));
  }

  public static async stairs1_up(): Promise<void> {
    await PSGame.warp(5, 7, false);
  }

  public static async stairs1_down(): Promise<void> {
    await PSGame.warp(21, 10, false);
  }

  public static async stairs2_up(): Promise<void> {
    await PSGame.warp(12, 12, false);
  }

  public static async stairs2_down(): Promise<void> {
    await PSGame.warp(25, 13, false);
  }

  public static async stairs3_up(): Promise<void> {
    await PSGame.warp(9, 3, false);
  }

  public static async stairs3_down(): Promise<void> {
    await PSGame.warp(25, 6, false);
  }

  public static async stairs4_up(): Promise<void> {
    await PSGame.warp(13, 7, false);
  }

  public static async stairs4_down(): Promise<void> {
    await PSGame.warp(27, 8, false);
  }

  public static async trap1(): Promise<void> {
    await PSGame.trapRoutine(Trap.GUARON_TRAP1, Trap.INFO_GUARON_TRAP1, 25, 8);
  }

  public static async trap2(): Promise<void> {
    await PSGame.trapRoutine(Trap.GUARON_TRAP2, Trap.INFO_GUARON_TRAP2, 21, 12);
  }

  public static async exit(): Promise<void> {
    await PSGame.mapswitchToPlanet(Planet.DEZORIS, 80, 7);
  }
}
